### DISCUSSION TITLE
in anyhow mode, how does the middleware obtain the anyhow err

### DISCUSSION THREAD

**Discussion by liaohongxing** [salvo-rs/salvo#82] [Q&A]
in anyhow mode, how does the middleware obtain the anyhow err

**Comment by chrislearn** [comment]
If you enable anyhow, it just let anyhow::Error implement ```Writer``` trait.

```
#![feature = "anyhow"]
#[async_trait]
impl Writer for anyhow::Error {
    #[inline]
    async fn write(self, _req: &mut Request, _depot: &mut Depot, res: &mut Response) {
        res.set_status_error(StatusError::internal_server_error());
    }
}
```

What is your specific requirement?

**Comment by liaohongxing** [comment]
我需要通过中间件或全局配置在控制台显示错误并且方便收集错误, 当然用tracing到处打点是一个办法 , 内置的LogHandler只显示500， 不显示anyhow返回的具体错误， 开发调试、错误收集都不方便

**Comment by chrislearn** [comment]
一般项目开发时, 你都需要自定义一个错误类型, 你让那个错误类型实现 Writer, 然后自己想怎么写都行. 
LogHandler 这种中间件本身也没几行代码, 复制修改一下也可以.

**Comment by chrislearn** [comment]
其实 anyhow 的错误其实也是被当成了 500 错误的. 如果想细粒度控制, 建议是自定义错误类型.
